### PR TITLE
speed up computing action masks

### PR DESCRIPTION
**Pull request by kywch**
These mask functions in `observation.py` were optimized (based on timeit 1000 iter)
* _make_move_mask(): 0.170 -> 0.022
* _make_attack_mask(): 0.060 -> 0.037
* _make_buy_mask(): 0.022 -> 0.011

Observation.tile() was expensive, so _make_move_mask directly access the tile map by computing the indices, and the tests were added to make sure the coordinates are correct. Also, expensive np.concatenate and redundant operations were removed from attack/buy masks.

Observation.tile() was also changed to compute indices directly and "parse array" internally.

### CODE REVIEW

**Comment by kywch** [comment]
config.IMMORTAL had a bug (fixed in the task-rev branch) and the agents were dying, so the numbers here would be hard to compare against the task-rev branch.

**Comment by kywch** [comment]
Tried this on the `task-rev` branch, where the IMMORTAL flag was fixed. 
* obs.to_gym() without action targets: 1.05 -> 0.30
* obs.to_gym() with action targets: 3.90 -> 2.44

This is the testing code:
```python
  config = nmmo.config.Default()
  config.PLAYERS = [Sleeper] # the scripted agents doing nothing
  config.IMMORTAL = True # otherwise the agents will die
  config.PROVIDE_ACTION_TARGETS = True
  env = nmmo.Env(config)
  env.reset(seed=0)
  for _ in range(3):
    env.step({})

  obs = env._compute_observations()

  test_func = [
    ('env.step({})', lambda: env.step({})),
    ('env.realm.step()', lambda: env.realm.step({})),
    ('env._compute_observations()', lambda: env._compute_observations()),
    ('obs.to_gym()', lambda: {a: o.to_gym() for a,o in obs.items()}),
    ('env._compute_rewards()', lambda: env._compute_rewards(obs.keys(), {}))
  ]

  for name, func in test_func:
    print(name, timeit(func, number=100, globals=globals()))
```

* `task-rev` without the mask patch, PROVIDE_ACTION_TARGETS = False
    env.step({}) 11.560627720999946
    env.realm.step() 3.808783102000234
    env._compute_observations() 3.317454328000167
    obs.to_gym() 1.0534005339995929
    env._compute_rewards() 2.7501322940006503

* `task-rev` without the mask patch, PROVIDE_ACTION_TARGETS = True
    env.step({}) 15.935197196000445  (??)
    env.realm.step() 3.776935551999486
    env._compute_observations() 3.3303598660004354
    obs.to_gym() 3.9035203909998017
    env._compute_rewards() 2.7542722939997475

* `task-rev` + the mask patch, PROVIDE_ACTION_TARGETS = True
    env.step({}) 12.784669539999413
    env.realm.step() 3.767843952000476
    env._compute_observations() 3.403790061000109
    obs.to_gym() 2.4430539750001117
    env._compute_rewards() 2.737189865999426

* `task-rev` + the mask patch, PROVIDE_ACTION_TARGETS = False
    env.step({}) 10.783011086999977
    env.realm.step() 3.8612988949998908
    env._compute_observations() 3.378951790999963
    obs.to_gym() 0.30076548800025193
    env._compute_rewards() 2.8621663109997826

File: nmmo/core/observation.py [hunk 1]
```diff
@@ -87,11 +88,13 @@ def tile(self, r_delta, c_delta):
...
     agent = self.agent()
+    center = self.config.PLAYER_VISION_RADIUS
+    tile_dim = self.config.PLAYER_VISION_DIAMETER
+    mat_map = self.tiles[:,2].reshape(tile_dim,tile_dim)
     if (0 <= agent.row + r_delta < self.config.MAP_SIZE) & \
        (0 <= agent.col + c_delta < self.config.MAP_SIZE):
-      r_cond = (self.tiles[:,TileState.State.attr_name_to_col["row"]] == agent.row + r_delta)
-      c_cond = (self.tiles[:,TileState.State.attr_name_to_col["col"]] == agent.col + c_delta)
-      return TileState.parse_array(self.tiles[r_cond & c_cond][0])
+      return SimpleNamespace(**{'row': agent.row+r_delta, 'col': agent.col+c_delta,
```
Comment (daveey) [inline nmmo/core/observation.py]: Are you saying it's faster to construct a SimpleNamespace instead of TileState?

**Review by daveey — APPROVED** [review]

File: nmmo/core/observation.py (hunk 1 above)
Comment (kywch) [reply nmmo/core/observation.py]: let me quickly test this

Comment (kywch) [reply nmmo/core/observation.py]: no difference, so I'm using parse_array again

**Comment by kywch** [comment]
_make_move_mask() was made faster by comparing the material id to `material.Habitable.indices` . Although the material id can be compared to material.Habitable, it goes through unnecessary expensive operation.  Having confirmed that the most of speed gain comes from the new tile, _make_move_mask() was reverted back with `material.Habitable.indices`.

Observation.tile() reverted back to using parse_array after confirming the cost was mainly calculating the array index.

An env.step() profile function was added to the testhelpers.

**Comment by kywch** [comment]
merging in to the main branch

**Pull request merged**